sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py

```diff
@@ -764,3 +760,188 @@ def update_repository_properties(self, *args, **kwargs):
+        )
+        digest = _compute_digest(data)
+        complete_upload_response = self._client.container_registry_blob.complete_upload(
+            digest=digest, next_link=upload_chunk_response['Location'], cls=_return_response_header, **kwargs
+        )
+        return complete_upload_response['Docker-Content-Digest']
+
+    @distributed_trace
+    def download_manifest(self, repository, tag_or_digest, **kwargs):
+        # type: (str, str, **Any) -> DownloadManifestResult
+        """Download the manifest for an OCI artifact.
+
+        :param str repository: Name of the repository
+        :param str tag_or_digest: The tag or digest of the manifest to download.
+        :returns: DownloadManifestResult
+        :rtype: ~azure.containerregistry.models.DownloadManifestResult
+        :raises ValueError: If the parameter repository or tag_or_digest is None.
+        :raises ~azure.core.exceptions.HttpResponseError:
+            If the requested digest does not match the digest of the received manifest.
+        """
+        if repository is None or tag_or_digest is None:
+            raise ValueError("The parameter repository and tag_or_digest cannot be None.")
+
+        response = self._client.container_registry.get_manifest(
+            name=repository,
+            reference=tag_or_digest,
+            accept=OCI_MANIFEST_MEDIA_TYPE,
+            headers={"Accept": OCI_MANIFEST_MEDIA_TYPE},
+            cls=_return_response,
+            **kwargs
+        )
+        digest = response.http_response.headers['Docker-Content-Digest']
+        content = response.http_response.internal_response._content
+        content_stream = BytesIO(content)
+        
+        if not _validate_digest(content_stream, digest):
+            raise HttpResponseError("The requested digest does not match the digest of the received manifest.")
+        
+        manifest = _deserialize_manifest(content_stream)
+        return DownloadManifestResult(digest=digest, data=content_stream, manifest=manifest)
+
+    @distributed_trace
+    def download_blob(self, repository, digest, **kwargs):
+        # type: (str, str, **Any) -> DownloadBlobResult | None
+        """Download a blob that is part of an artifact.
+
+        :param str repository: Name of the repository
+        :param str digest: The digest of the blob to download.
+        :returns: DownloadBlobResult or None
+        :rtype: ~azure.containerregistry.DownloadBlobResult or None
+        :raises ValueError: If the parameter repository or digest is None.
+        """
+        if repository is None or digest is None:
+            raise ValueError("The parameter repository and digest cannot be None.")
+
+        result = self._client.container_registry_blob.get_blob(repository, digest, **kwargs)
+        
+        if not result:
+            return result
+        
+        data = BytesIO(result.response.internal_response.content)
+        digest = _compute_digest(data)
```

Comment:
Is this going to read the entire result stream? Or is this data not actually 'streamed' from the service?
That's possible going to cause other problems - it seems inefficient to 'download' the data twice - how big could it possible get?
Additionally, the connection will be closed after the stream has been read, so unless we're saving that data in memory it will be lost.
Is the digest not returned in a header we could use?

Reply:
- Yes, it is reading the entire result stream in _compute_digest().
- The deserialized data returned from service is in type of StreamDownloadGenerator
- The digest is not returned in headers

Reply:
Okay - we probably shouldn't block the PR on this one right now - but we will need to dig into this more closely before GA.
It sounds like there's no real limits, expectations around data size and I'm a bit concerned that we're making decisions based on small data sizes that are going to be inefficient or difficult to use if the payload is ever larger. We can open an issue for this for now and big in more deeply before GA.